platform/android/src/style/sources/geojson_source.cpp

```diff
@@ -86,6 +154,39 @@ namespace android {
         source.as<mbgl::style::GeoJSONSource>()->GeoJSONSource::setGeoJSON(GeoJSON(feature));
     }
 
+    void GeoJSONSource::setFeatureAsync(JNIEnv& env, jni::Object<geojson::Feature> jFeature,
+                                   jni::Object<OnGeoJsonSourceLoadedListener> jListener) {
+        // If another update is running, log an error
+        if (callback) {
+            mbgl::Log::Error(mbgl::Event::JNI, "Error setting GeoJSON: another asynchronous update of this source is being processed.");
+            return;
+        }
+
+        callback = std::make_unique<Actor<FeatureConverter::Callback>>(
+                *Scheduler::GetCurrent(),
+                [this](GeoJSON geoJSON) {
+                    android::UniqueEnv _env = android::AttachEnv();
+
+                    // Update the core source
+                    source.as<mbgl::style::GeoJSONSource>()->GeoJSONSource::setGeoJSON(geoJSON);
+
+                    featureRef.release();
+                    callback.release();
+                    callback = nullptr;
+
+                    // Obtain a local reference to the listener and release the global reference
+                    jni::Object<OnGeoJsonSourceLoadedListener> listener = *sourceLoadedListenerRef;
+                    sourceLoadedListenerRef.release();
+
+                    // Notify the update listener
+                    OnGeoJsonSourceLoadedListener::notifySourceLoaded(*_env, listener);
+                });
+
+        featureRef = jFeature.NewGlobalRef(env);
```

Comment:
Let's see, the reason the `featureRef` has to live until the callback is really just that it has to live until `convertFeature` is done, but we can't free the `UniqueObject` from a different thread than the one it was created, right?

Instead of making `featureRef` etc. members of `GeoJSONSource` and manually managing the state, one thing you might try is to make a shared_ptr to the `UniqueObject`, and then capture it in the callback lambda. The lambda will hold onto a copy of the shared_ptr, and then when the callback finishes it will free the resource without you having to do any further management.

So something like:

```
std::shared_ptr<jni::UniqueObject<type>> featureRef = jFeature.NewGlobalRef(env);
callback = ...
  [this, featureRef] {
    ...  do unrelated stuff
    featureRef.reset(); // This would actually happen implicitly at the end of the block, but since
                        // you're not otherwise using featureRef it's nice to be explicit
}

Reply:
> but we can't free the UniqueObject from a different thread than the one it was created, right?

That's a great question I don't know the answer to, but anyway, capturing the ref in the lambda looks clean, thanks!